File: sample_app/first_app/report/workflow_status_report/workflow_status_report.js
// Copyright (c) 2016, Leader Group and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Workflow Status Report"] = {
	"filters": [
		{
			"fieldname":"comment_type",
			"label": __("Comment Type"),
			"fieldtype": "Data",
            "default": "Workflow",
            "hidden": 1
		}, 
		{
			"fieldname":"reference_doctype",
			"label": __("Reference Type"),
			"fieldtype": "Select",
			"options": [
				"",
				{
					label: __("Sales Order"),
					value: "Sales Order",
				},
				{
					label: __("Purchase Order"),
					value: "Purchase Order",
				},
				{
					label: __("Material Request"),
					value: "Material Request",
				},
				{
					label: __("Payment Request"),
					value: "Payment Request",
				},
				{
					label: __("Delivery Note"),
					value: "Delivery Note",
				},
				{
					label: __("Stock Entry"),
					value: "Stock Entry",
				},
				{
					label: __("Purchase Receipt"),
					value: "Purchase Receipt",
				},
			],
			"reqd": 1,
            "default": "Sales Order",
            // on_change: function() {
            //     frappe.query_report.set_filter_value('reference_name', "");
            //                 frappe.query_report.refresh();
            // },
		},
		{
			"fieldname":"reference_name",
			"label": __("Reference Document"),
            // "fieldtype": "MultiSelectList",
            "fieldtype": "Data",
            "hidden": 1
            // get_data: function(txt) {
            //                 let ref_doc = frappe.query_report.get_filter_value('reference_doctype');
            //     return frappe.db.get_link_options(ref_doc, txt);
            // },
		}  
	]
};

File: sample_app/first_app/report/workflow_status_report/workflow_status_report.py
# Copyright (c) 2013, Leader Group and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import formatdate

def execute(filters=None):
    columns, data = [], []
    columns = get_columns()

    # if not filters.get("reference_name"):
    #     del filters["reference_name"]
    # else:
    #     reference_name = filters.get("reference_name")[0]
    #     filters.reference_name = reference_name

    data = frappe.db.get_all("Comment",
            filters=filters,
            fields=["owner",
                    "reference_name",
                    "creation",
                    'content'
                ],
            order_by="creation")
    for d in data:
        d['creation'] = formatdate(d['creation'])
        d['owner'] =  frappe.get_value('User', d.owner, 'full_name')
        workflow = frappe.get_value('Workflow', {'document_type': filters["reference_doctype"], 'is_active': 1}, 'name')
        transition = frappe.get_value('Workflow Transition', {'parent': workflow, 'next_state': d.content}, ['state', 'action'], as_dict=1)
        if transition:
            d['previous_transition'] = transition.state
            d['action'] = transition.action
    return columns, data

def get_columns():
    return [
        {
            "fieldname": "reference_name",
            "label": _("Reference Name"),
            "fieldtype": "Data",
            "width": 180
        },
        {
            "fieldname": "owner",
            "label": _("Approved By"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "creation",
            "label": _("Approved On"),
            "fieldtype": "DateTime",
            "width": 180
        },
        {
            "fieldname": "previous_transition",
            "label": _("From State"),
            "fieldtype": "Data",
            "width": 300
        },
        {
            "fieldname": "action",
            "label": _("Action"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "content",
            "label": _("To State"),
            "fieldtype": "Data",
            "width": 300
        },
    ]